### PR TITLE
feat(riklet): implement workload function deletion

### LINKED ISSUE #88: Correct firecracker process management
## What needs to be done ?

- `firepilot` VM stop method isn't properly killing the process of the VM (process stays alive)
- `firepilot` VM start is blocking, it should be running in a separate process handled by `firepilot`
- `firepilot` start of VM should return an object that can manage the VM, it shouldn't return a String
- `firepilot` uses a complete configuration and process it as a JSON, we should be using the API of the `firecracker` socket

We got a few solution for this, we can either continuing developing `firepilot` and improve it to match our needs, or, we could go for a different crate like [`firec`](https://github.com/blockjoy/firec).

## Why it needs to be done ?

- We can hardly manage the mVM
- We cannot configure it beforehand and then run it in a second step (e.g create TAP, configure IP and then run the mVM)
- We cannot properly stop the VM

### PR DESCRIPTION
**Pull request by AlexandreBrg**
Closes #88 

## 📑 Description

* Add ability to delete a workload of type function
* Add ability to re-scheduler a workload instance that has been previously delted
* Migrate to `firepilot` v1.0.0
* Add some network related documentation

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code is documented
- [ ] I provided unitary tests or procedure to test my code
- [ ] My PR have a clear description of what my code is supposed to do

### CODE REVIEW

File: riklet/Cargo.toml [hunk 1]
```diff
@@ -41,7 +41,7 @@ uuid = { version = "1.3", features = ["v4"] }
 clap = { version = "4.0.26", features = ["derive", "env"] }
 nix="0.21.0"
 gethostname = "0.4.1"
-firepilot = { git = "https://github.com/rik-org/firepilot.git", tag = "v0.2.0" }
+firepilot = "1.1.0"
```
Comment (MaloPolese) [inline riklet/Cargo.toml:44]: Nice !

File: riklet/src/net_utils.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+use futures_util::TryStreamExt;
+use rand::Rng;
+use rtnetlink::new_connection;
+use tracing::{trace, warn};
+use utils::net::mac::MacAddr;
+
+#[tracing::instrument()]
+pub async fn set_link_up(iface_name: String) -> Result<(), rtnetlink::Error> {
+    trace!("link {} up", &iface_name);
+    let (connection, handle, _) = new_connection().unwrap();
```
Comment (MaloPolese) [inline riklet/src/net_utils.rs:12]: 🟠 I don't think this unwrap is safe

File: riklet/src/net_utils.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+#[tracing::instrument()]
+pub async fn set_link_ipv4(
+    iface_name: String,
+    ipv4: Ipv4Addr,
+    mask: u8,
+) -> Result<(), rtnetlink::Error> {
+    trace!("link {} ipv4: {}/{}", &iface_name, ipv4, mask);
+    let (connection, handle, _) = new_connection().unwrap();
```
Comment (MaloPolese) [inline riklet/src/net_utils.rs:33]: 🟠 same

File: riklet/src/net_utils.rs (hunk 2 above)
Comment (AlexandreBrg) [reply riklet/src/net_utils.rs:12]: It's fairly safe, as it won't be able to unwrap only if `rtnetlink` is not available as a socket. That said, linux kernel uses rtnetlink for routing ([docs](https://man7.org/linux/man-pages/man7/rtnetlink.7.html)). If the user doesn't have rtnetlink, it means he doesn't have Linux, and we don't support MacOS / Linux

File: riklet/src/net_utils.rs (hunk 3 above)
Comment (AlexandreBrg) [reply riklet/src/net_utils.rs:33]: Ditto
https://github.com/rik-org/rik/pull/94#discussion_r1175139793

File: riklet/src/runtime/function_runtime.rs [hunk 4]
```diff
@@ -74,56 +99,54 @@ impl FunctionRuntime {
...
     #[tracing::instrument(skip(self))]
-    async fn down(&self) -> Result<()> {
+    async fn down(&mut self) -> Result<()> {
         debug!("Destroying function runtime vm");
-        if let Some(vm_config) = &self.vm_config {
-            let firecracker = Firecracker::new(firepilot::FirecrackerOptions {
-                command: Some(self.function_config.firecracker_location.clone()),
-                ..Default::default()
-            })
+        let machine = self.machine.as_mut().unwrap();
```
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply riklet/src/runtime/function_runtime.rs:137]: Fixed

File: scheduler/src/state_manager/mod.rs [hunk 5]
```diff
@@ -414,7 +412,7 @@ impl StateManager {
             self.action_minus_replicas(&request.workload_id, def_replicas)?;
         } else {
             info!("Workload is getting unscheduled");
-            workload.status = ResourceStatus::Destroying;
+            // workload.status = ResourceStatus::Destroying;
```
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply scheduler/src/state_manager/mod.rs:415]: Done

File: riklet/src/runtime/function_runtime.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply riklet/src/runtime/function_runtime.rs:137]: Handled

File: riklet/src/net_utils.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
+use std::net::Ipv4Addr;
+
+use futures_util::TryStreamExt;
+use rand::Rng;
+use rtnetlink::new_connection;
+use tracing::{trace, warn};
+use utils::net::mac::MacAddr;
+
+#[tracing::instrument()]
+pub async fn set_link_up(iface_name: String) -> Result<(), rtnetlink::Error> {
```
[the remark this answers is not in the document]
Comment (MaloPolese) [reply riklet/src/net_utils.rs:10]: I agree

**Review by MaloPolese — APPROVED** [review]

**Review by kalil-pelissier — APPROVED** [review]

**Pull request merged**